README.md
A simple and beautiful way to display a four sided widget with a basic 3D effect.

## Usage

The simplest way is to use the ``RotationStage`` widget.
You only have to provide a ``contentBuilder``, everything else is preconfigured.

```dart
Widget build(BuildContext context) {
  return RotationStage(
    contentBuilder: (int index,
        RotationStageSide side,
        double currentPage,) =>
        Card(
          child: Padding(
            padding: const EdgeInsets.all(8.0),
            child: Text(
              side.map(
                front: "Front",
                left: "Left",
                back: "Back",
                right: "Right",
              ),
            ),
          ),
        ),
  );
}
```

You can rotate the widget by swiping on the bottom bar. The top part is purposfully not swipeable,
so you can listen to whatever gestures you want there.

If you want more fine-grained control, check out the other parameters of the constructor, or
``RotationStageBar``, ``RotationStageHandle`` and ``RotationStageContent``.

The source code for ``RotationStage`` should be a good starting point.

## Example

To run the example open the ``example`` folder and run ``flutter create .``

CHANGELOG.md
## 0.1.0
* Adjusted chips for Flutter 3.3
## 0.0.9
* added center to bar

## 0.0.8
* removed unnecessary center from handle

## 0.0.7
* smaller start page to prevent exception

## 0.0.6
* lint dependency

## 0.0.5
* Fixed handle color

## 0.0.4
* Fixed handle color
* Allow custom handle colors

## 0.0.3
* Added support for localization or custom labels via inherited widget
## 0.0.2

* BREAKING: Removed initial page parameter from controller constructor

## 0.0.1

* Initial Release, check README for details


lib/rotation_stage.dart
library rotation_stage;

import 'package:flutter/material.dart';
import 'package:rotation_stage/rotation_stage.dart';

export 'src/model/rotation_stage_side.dart';
export 'src/rotation_stage_bar.dart';
export 'src/rotation_stage_content.dart';
export 'src/rotation_stage_controller.dart';
export 'src/rotation_stage_handle.dart';
export 'src/rotation_stage_labels.dart';

typedef RotationStageBuilder = Widget Function(
    int index, RotationStageSide side, double currentPage);

class RotationStage extends StatefulWidget {
  const RotationStage({
    super.key,
    required this.contentBuilder,
    this.controller,
    this.viewHandleBuilder,
    this.labels,
    this.barHeight = 64,
    this.barInteractable = true,
  });

  final RotationStageBuilder contentBuilder;
  final RotationStageController? controller;
  final RotationStageBuilder? viewHandleBuilder;
  final RotationStageLabelData? labels;
  final double barHeight;
  final bool barInteractable;

  @override
  State<RotationStage> createState() => _RotationStageState();
}

class _RotationStageState extends State<RotationStage> {
  late final RotationStageController _controller;

  @override
  void initState() {
    _controller = widget.controller ?? RotationStageController();
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return RotationStageLabels(
      data: widget.labels ?? RotationStageLabelData.english,
      child: Column(
        mainAxisSize: MainAxisSize.min,
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          Expanded(
            child: RotationStageContent(
              controller: _controller,
              contentBuilder: widget.contentBuilder,
            ),
          ),
          const Divider(
            height: 1,
          ),
          RotationStageBar(
            controller: _controller,
            interactable: widget.barInteractable,
            viewHandleBuilder: widget.viewHandleBuilder ??
                (index, side, page) => RotationStageHandle(
                      onTap: () => _controller.animateToPage(index),
                      side: side,
                      active: index == page.round(),
                      backgroundTransparent: !widget.barInteractable,
                    ),
          ),
        ],
      ),
    );
  }
}


lib/src/rotation_stage_bar.dart
import 'dart:ui';

import 'package:flutter/material.dart';
import 'package:rotation_stage/src/rotation_stage_controller.dart';

import '../rotation_stage.dart';

class RotationStageBar extends StatelessWidget {
  const RotationStageBar({
    super.key,
    required this.controller,
    required this.viewHandleBuilder,
    this.height = kToolbarHeight,
    this.interactable = true,
    this.minHandleOpacity = 0,
  })  : assert(minHandleOpacity >= 0),
        assert(minHandleOpacity <= 1);

  final RotationStageController controller;
  final RotationStageBuilder viewHandleBuilder;
  final bool interactable;
  final double height;
  final double minHandleOpacity;

  @override
  Widget build(BuildContext context) {
    final visOffset = 0.5 / controller.pageController.viewportFraction;
    return SizedBox(
      height: height,
      child: ValueListenableBuilder<double>(
        valueListenable: controller,
        builder: (context, page, _) => PageView.builder(
          controller: controller.pageController,
          itemBuilder: (context, index) {
            final offset = (page - index).abs().clamp(0, visOffset) / visOffset;
            final opacity = lerpDouble(minHandleOpacity, 1, 1 - offset);
            return Center(
              child: Opacity(
                opacity: Curves.ease.transform(opacity!),
                child: AnimatedOpacity(
                  duration: kThemeAnimationDuration,
                  opacity: interactable && index != index ? 0 : 1,
                  child: viewHandleBuilder(
                    index,
                    RotationStageSide.forIndex(index),
                    page,
                  ),
                ),
              ),
            );
          },
        ),
      ),
    );
  }
}


lib/src/rotation_stage_handle.dart
import 'package:flutter/material.dart';
import 'package:rotation_stage/src/model/rotation_stage_side.dart';
import 'package:rotation_stage/src/rotation_stage_labels.dart';

class RotationStageHandle extends StatelessWidget {
  const RotationStageHandle({
    super.key,
    required this.side,
    required this.active,
    required this.onTap,
    required this.backgroundTransparent,
    this.activeForegroundColor,
    this.inactiveForegroundColor,
    this.activeBackgroundColor,
    this.inactiveBackgroundColor,
  });

  final RotationStageSide side;
  final bool active;
  final bool backgroundTransparent;
  final VoidCallback onTap;

  final Color? activeForegroundColor;
  final Color? inactiveForegroundColor;
  final Color? activeBackgroundColor;
  final Color? inactiveBackgroundColor;

  @override
  Widget build(BuildContext context) {
    final colorScheme = Theme.of(context).colorScheme;
    final labels = RotationStageLabels.of(context);
    final name = labels.getForSide(side);
    return RawChip(
      showCheckmark: false,
      onSelected: (_) => onTap(),
      label: Text(
        name.toUpperCase(),
        style: Theme.of(context).textTheme.button?.copyWith(
              color: active
                  ? activeForegroundColor ?? colorScheme.onPrimary
                  : inactiveForegroundColor ?? colorScheme.onPrimaryContainer,
            ),
      ),
      selected: active,
      disabledColor: Colors.transparent,
      shadowColor:
          backgroundTransparent ? Colors.transparent : colorScheme.shadow,
      selectedShadowColor: backgroundTransparent
          ? Colors.transparent
          : activeBackgroundColor ?? colorScheme.primary,
      backgroundColor: backgroundTransparent
          ? Colors.transparent
          : inactiveBackgroundColor ?? colorScheme.primaryContainer,
      selectedColor: backgroundTransparent
          ? Colors.transparent
          : activeBackgroundColor ?? colorScheme.primary,
    );
  }
}


lib/src/rotation_stage_labels.dart
import 'package:flutter/widgets.dart';
import 'package:rotation_stage/rotation_stage.dart';

class RotationStageLabelData {
  const RotationStageLabelData({
    required this.front,
    required this.left,
    required this.right,
    required this.back,
  });

  static const english = RotationStageLabelData(
    front: "Front",
    left: "Left",
    right: "Right",
    back: "Back",
  );

  final String front;
  final String left;
  final String right;
  final String back;

  String getForSide(RotationStageSide side) => side.map(
        front: front,
        left: left,
        back: back,
        right: right,
      );
}

class RotationStageLabels extends InheritedWidget {
  const RotationStageLabels({
    super.key,
    required this.data,
    required super.child,
  });

  final RotationStageLabelData data;

  static RotationStageLabelData of(context) {
    final RotationStageLabels? result =
        context.dependOnInheritedWidgetOfExactType<RotationStageLabels>();
    return result?.data ?? RotationStageLabelData.english;
  }

  @override
  bool updateShouldNotify(RotationStageLabels oldWidget) {
    return oldWidget.data != data;
  }
}


lib/src/rotation_stage_controller.dart
import 'package:flutter/material.dart';
import 'package:rotation_stage/src/const/scroll_start_page.dart';

class RotationStageController extends ValueNotifier<double> {
  RotationStageController({
    this.viewportFraction = 0.2,
  })  : pageController = PageController(
          initialPage: kInfiniteScrollStartPage,
          viewportFraction: viewportFraction,
        ),
        super(kInfiniteScrollStartPage.toDouble()) {
    pageController.addListener(() {
      if (pageController.positions.isNotEmpty && pageController.page != null) {
        value = pageController.page!;
      }
    });
  }

  final double viewportFraction;
  final PageController pageController;

  void animateToPage(
    int page, {
    Duration duration = kThemeAnimationDuration,
    Curve curve = Curves.ease,
  }) {
    pageController.animateToPage(
      page,
      duration: duration,
      curve: curve,
    );
  }

  @override
  void dispose() {
    pageController.dispose();
    super.dispose();
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:rotation_stage/rotation_stage.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Rotation Stage Example',
      theme: ThemeData(
        useMaterial3: true,
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.blue),
      ),
      home: const MyHomePage(title: 'Rotation Stage'),
      localizationsDelegates: [],
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: SafeArea(
        child: RotationStage(
          contentBuilder: (
            int index,
            RotationStageSide side,
            double currentPage,
          ) =>
              Padding(
            padding: const EdgeInsets.all(64),
            child: SizedBox.expand(
              child: Card(
                elevation: 0,
                color: Theme.of(context).colorScheme.inverseSurface,
                child: Center(
                  child: Card(
                    color: Colors.white,
                    child: Padding(
                      padding: const EdgeInsets.all(8.0),
                      child: Text(
                        side.map(
                          front: "Front",
                          left: "Left",
                          back: "Back",
                          right: "Right",
                        ),
                      ),
                    ),
                  ),
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }
}
